optaplanner-core/src/main/java/org/optaplanner/core/impl/score/stream/drools/bi/DroolsJoinBiConstraintStream.java

```diff
@@ -0,0 +1,127 @@
+package org.optaplanner.core.impl.score.stream.drools.bi;
+
+import java.util.Collection;
+
+import org.drools.model.Declaration;
+import org.drools.model.PatternDSL;
+import org.optaplanner.core.api.score.stream.bi.BiJoiner;
+import org.optaplanner.core.impl.score.stream.bi.AbstractBiJoiner;
+import org.optaplanner.core.impl.score.stream.common.JoinerType;
+import org.optaplanner.core.impl.score.stream.drools.DroolsConstraintFactory;
+import org.optaplanner.core.impl.score.stream.drools.uni.DroolsAbstractUniConstraintStream;
+
+public class DroolsJoinBiConstraintStream<Solution_, A, B> extends DroolsAbstractBiConstraintStream<Solution_, A, B> {
+
+    private final DroolsAbstractUniConstraintStream<Solution_, A> leftParentStream;
+    private final DroolsAbstractUniConstraintStream<Solution_, B> rightParentStream;
+    private final AbstractBiJoiner<A, B> biJoiner;
+    private final PatternDSL.PatternDef<B> rightPattern;
+
+    public DroolsJoinBiConstraintStream(DroolsConstraintFactory<Solution_> constraintFactory,
+            DroolsAbstractUniConstraintStream<Solution_, A> parent,
+            DroolsAbstractUniConstraintStream<Solution_, B> otherStream, BiJoiner<A, B> biJoiner) {
+        super(constraintFactory, null);
+        this.leftParentStream = parent;
+        this.rightParentStream = otherStream;
+        this.biJoiner = (AbstractBiJoiner<A, B>) biJoiner;
+        this.rightPattern = otherStream.getPattern().expr(getLeftVariableDeclaration(), (b, a) -> matches(a, b));
+    }
+
+    @Override
+    public DroolsAbstractUniConstraintStream<Solution_, A> getLeftParentStream() {
+        return leftParentStream;
+    }
+
+    @Override
+    public DroolsAbstractUniConstraintStream<Solution_, B> getRightParentStream() {
+        return rightParentStream;
+    }
+
+    @Override
+    public Declaration<A> getLeftVariableDeclaration() {
+        return leftParentStream.getVariableDeclaration();
+    }
+
+    @Override
+    public PatternDSL.PatternDef<A> getLeftPattern() {
+        return leftParentStream.getPattern();
+    }
+
+    @Override
+    public Declaration<B> getRightVariableDeclaration() {
+        return rightParentStream.getVariableDeclaration();
+    }
+
+    @Override
+    public PatternDSL.PatternDef<B> getRightPattern() {
+        return rightPattern;
+    }
+
+    private static boolean matches(Object left, Object right, JoinerType type) {
```

Comment:
Suggestion: how about  putting the matching logic into the `JoinerType` enum? Is it going to be reused for `TriStreams` too?

Another option is a hierarchy of matcher classes per JoinerType.

Reply:
I'm going to move the code if I need to reuse it. Right now, I didn't really think about `TriStream`s, but the code will definitely evolve as needed.